DatadogRUM/Sources/RUMMetrics/ITNVMetric.swift

```diff
@@ -0,0 +1,111 @@
+    /// - Parameters:
+    ///   - viewID: The ID of the view for which the ITNV value is requested.
+    /// - Returns: The ITNV value (time interval) for the specified view, or `nil` if not available (the view has been completed).
+    func value(for viewID: RUMUUID) -> TimeInterval?
+}
+
+internal final class ITNVMetric: ITNVMetricTracking {
+    enum Constants {
+        /// The maximum allowed duration for the ITNV metric. Values exceeding this threshold are ignored.
+        static let maxDuration: TimeInterval = 3
+    }
+
+    /// The time of the last recorded action in the previous view.
+    private var lastActionDateByViewID: [RUMUUID: Date] = [:]
+
+    /// Stores the start times of views.
+    private var startDateByViewID: [RUMUUID: Date] = [:]
+
+    /// Tracks the previous view associated with each view.
+    private var previousViewByViewID: [RUMUUID: RUMUUID] = [:]
+
+    /// The ID of the current view.
+    private var currentViewID: RUMUUID?
+
+    func trackAction(startTime: Date, endTime: Date, actionType: RUMActionType, in viewID: RUMUUID) {
+        // Retrieve the last recorded action time for the given view
+        let lastDate = lastActionDateByViewID[viewID]
+        switch actionType {
+        case .tap, .click: // Discrete actions like tap or click should use their start time
+            lastActionDateByViewID[viewID] = max(startTime, lastDate ?? .distantPast)
+        case .swipe: // Continuous actions like swipe should use their end time
+            lastActionDateByViewID[viewID] = max(endTime, lastDate ?? .distantPast)
+        case .scroll, .custom:
+            return // Ignore scroll and custom actions for ITNV calculation
+        }
+    }
+
+    func trackViewStart(at viewStart: Date, viewID: RUMUUID) {
+        startDateByViewID[viewID] = viewStart
+        previousViewByViewID[viewID] = currentViewID
+        currentViewID = viewID
+    }
+
+    func trackViewComplete(viewID: RUMUUID) {
+        startDateByViewID[viewID] = nil
+
+        if let previousViewID = previousViewByViewID[viewID] {
+            lastActionDateByViewID[previousViewID] = nil
+        }
+        previousViewByViewID[viewID] = nil
+
+        if viewID == currentViewID {
+            currentViewID = nil
+        }
+    }
+
+    func value(for viewID: RUMUUID) -> TimeInterval? {
+        guard let viewStartDate = startDateByViewID[viewID] else {
+            return nil // View has not started yet
+        }
+
+        guard let previousViewID = previousViewByViewID[viewID] else {
+            return nil // No previous view for this one
+        }
+
+        guard let lastActionDate = lastActionDateByViewID[previousViewID] else {
+            return nil // No action recorded in the previous view
+        }
+
+        let itnvValue = viewStartDate.timeIntervalSince(lastActionDate)
+
+        guard itnvValue <= Constants.maxDuration else {
+            return nil // ITNV exceeds the maximum allowed duration, return nil
```

Comment:
👍 This will be covered in `RUM-7111` where we will design RUM telemetry to track the performance of this metric. That will include both distribution of values (health-check) and common reasons for reaching `nil` value.